From 850b55dc4697cfe160652deb193eb9e467333386 Mon Sep 17 00:00:00 2001
From: danielhk <daniel.p6800@gmail.com>
Date: Sun, 3 Dec 2017 22:52:31 +0800
Subject: [PATCH] patch: frameworks_av_libaudiohal

Oreo's audio_stream_in_t definition changed. Old blobs from MTK (and
others) is not compatible. This patch would hopefully work for all old
audio HALs and fix all the issues arised.
1. At openInputStream(), store a copy of the struct from audio HAL to
mIn.
2. Assign the supported entries to halStream and assign NULL to those
unsupported.
3. close the mIn and free the memory at closeOutputStream()
Note: legacy_audio_stream_in_t is defined in the overrided audio.h
---
 media/libaudiohal/DeviceHalLocal.cpp | 18 ++++++++++++++++++
 media/libaudiohal/DeviceHalLocal.h   |  4 +++-
 2 files changed, 21 insertions(+), 1 deletion(-)

diff --git a/media/libaudiohal/DeviceHalLocal.cpp b/media/libaudiohal/DeviceHalLocal.cpp
index fc098f513..006faa0dd 100644
--- a/media/libaudiohal/DeviceHalLocal.cpp
+++ b/media/libaudiohal/DeviceHalLocal.cpp
@@ -137,8 +137,21 @@ status_t DeviceHalLocal::openInputStream(
             handle, devices, flags,
             config->sample_rate, config->format, config->channel_mask,
             address, source);
+#ifdef MTK_HARDWARE
+    halStream = (audio_stream_in_t *)calloc(1, sizeof(audio_stream_in_t));
+    int openResult = mDev->open_input_stream(
+            mDev, handle, devices, config, &mIn, flags, address, source);
+    memcpy(halStream, mIn, sizeof(legacy_audio_stream_in_t));
+    halStream->get_capture_position = NULL;
+    halStream->start = NULL;
+    halStream->stop = NULL;
+    halStream->create_mmap_buffer = NULL;
+    halStream->get_mmap_position = NULL;
+#else
     int openResult = mDev->open_input_stream(
             mDev, handle, devices, config, &halStream, flags, address, source);
+#endif
+
     if (openResult == OK) {
         *inStream = new StreamInHalLocal(halStream, this);
     }
@@ -193,7 +206,12 @@ void DeviceHalLocal::closeOutputStream(struct audio_stream_out *stream_out) {
 }
 
 void DeviceHalLocal::closeInputStream(struct audio_stream_in *stream_in) {
+#ifdef MTK_HARDWARE
+    mDev->close_input_stream(mDev, mIn);
+    free(stream_in);
+#else
     mDev->close_input_stream(mDev, stream_in);
+#endif
 }
 
 } // namespace android
diff --git a/media/libaudiohal/DeviceHalLocal.h b/media/libaudiohal/DeviceHalLocal.h
index 865f2968e..514b12e18 100644
--- a/media/libaudiohal/DeviceHalLocal.h
+++ b/media/libaudiohal/DeviceHalLocal.h
@@ -107,7 +107,9 @@ class DeviceHalLocal : public DeviceHalInterface
 
   private:
     audio_hw_device_t *mDev;
-
+#ifdef MTK_HARDWARE
+    audio_stream_in_t *mIn;
+#endif
     friend class DevicesFactoryHalLocal;
 
     // Can not be constructed directly by clients.
-- 
2.11.0

